plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/RedshiftUnloadSplitSource.java

```diff
@@ -0,0 +1,153 @@
+                log.info("UNLOAD command for %s query took %sms", queryFragmentId, System.currentTimeMillis() - beginTs);
+            }
+            catch (SQLException e) {
+                if (e instanceof RedshiftException && e.getMessage().contains("The S3 bucket addressed by the query is in a different region from this cluster")) {
+                    throw new TrinoException(REDSHIFT_S3_CROSS_REGION_UNSUPPORTED, "Redshift cluster and S3 bucket in different regions is not supported", e);
+                }
+                throw new RuntimeException(e);
+            }
+        }, executor);
+    }
+
+    @Override
+    public CompletableFuture<ConnectorSplitBatch> getNextBatch(int maxSize)
+    {
+        return resultSetFuture
+                .thenApply(_ -> {
+                    ConnectorSplitBatch connectorSplitBatch = new ConnectorSplitBatch(getUnloadedFilePaths().stream()
+                            .map(fileInfo -> (ConnectorSplit) new RedshiftUnloadSplit(fileInfo.path, fileInfo.size))
+                            .collect(toImmutableList()), true);
+                    finished = true;
+                    return connectorSplitBatch;
+                });
+    }
+
+    private List<FileInfo> getUnloadedFilePaths()
+    {
+        manifestLocation = Optional.of(Location.of(unloadOutputPath + "manifest"));
+        TrinoInputFile inputFile = fileSystem.newInputFile(manifestLocation.get());
+        JsonNode outputFileEntries;
+        try (TrinoInputStream inputStream = inputFile.newStream()) {
+            byte[] manifestContent = inputStream.readAllBytes();
+            outputFileEntries = OBJECT_MAPPER.readTree(manifestContent).path("entries");
+        }
+        // Rely on the catching `FileNotFoundException` for the absence of manifest file as `inputFile#exists` adds additional call to S3.
+        catch (FileNotFoundException e) {
+            // manifest is not generated if unload query doesn't produce any results.
+            manifestLocation = Optional.empty();
+            return ImmutableList.of();
+        }
+        catch (IOException e) {
+            throw new TrinoException(REDSHIFT_FILESYSTEM_ERROR, e);
+        }
+        ImmutableList.Builder<FileInfo> unloadedFilePaths = ImmutableList.builder();
+        outputFileEntries.elements()
+                .forEachRemaining(fileInfo -> unloadedFilePaths.add(new FileInfo(fileInfo.get("url").asText(), fileInfo.get("meta").get("content_length").longValue())));
+        return unloadedFilePaths.build();
+    }
+
+    @Override
+    public void close()
+    {
+        resultSetFuture.cancel(true);
+        manifestLocation.ifPresent(location -> {
+            try {
+                fileSystem.deleteFile(location);
+            }
+            catch (IOException e) {
+                log.info("Failed to delete manifest file: " + location);
+            }
+        });
+        try {
+            statement.close();
+            connection.close();
+        }
```

Comment:
We shouldn't have to wait until end of split source to close the connection and statement, i think it can be done as soon as the UNLOAD has finished